argonaut-common/man/argonaut.conf.1
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ARGONAUT.CONF 1"
.TH ARGONAUT.CONF 1 "2019-05-20" "Argonaut 1.3.1" "Argonaut Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
argonaut.conf \- Argonaut configuration file
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The argonaut.conf file is used to configure the argonaut server and client
.SS "Configuration layout"
.IX Subsection "Configuration layout"
The configuration is a classical ini style file
.SS "Server options"
.IX Subsection "Server options"
.IP "\fBserver_ip\fR \fIip address\fR" 4
.IX Item "server_ip ip address"
The server_ip statement give the ip of the argonaut server
.SS "Client section options"
.IX Subsection "Client section options"
.IP "\fBclient_ip\fR \fIip address\fR" 4
.IX Item "client_ip ip address"
The client_ip statement give the ip of the argonaut server
.SS "\s-1LDAP\s0 options"
.IX Subsection "LDAP options"
.IP "\fBconfig\fR \fIpath\fR" 4
.IX Item "config path"
Path to the file that contains the ldap \s-1URI\s0 and base
.IP "\fBdn\fR \fIdn\fR" 4
.IX Item "dn dn"
Complete ldap \s-1DN\s0 of the \s-1DSA\s0 account needed to read the argonaut attributes from the ldap server
.IP "\fBpassword\fR \fIstring\fR" 4
.IX Item "password string"
Password of the \s-1DSA\s0 account
.IP "\fBtls\fR \fIbool\fR" 4
.IX Item "tls bool"
By default to off, switch to on if you want to use ssl on argonaut
.SH "BUGS"
.IX Header "BUGS"
Please report any bugs, or post any suggestions, to the fusiondirectory mailing list fusiondirectory-users or to <https://gitlab.fusiondirectory.org/argonaut/argonaut/issues/new>
.SH "LICENCE AND COPYRIGHT"
.IX Header "LICENCE AND COPYRIGHT"
This code is part of Argonaut Project <https://www.argonaut\-project.org>
.IP "Copyright (C) 2011\-2018 FusionDirectory project" 2
.IX Item "Copyright (C) 2011-2018 FusionDirectory project"
.PP
This program is distributed in the hope that it will be useful, but \s-1WITHOUT ANY WARRANTY\s0; without even the implied warranty of \s-1MERCHANTABILITY\s0 or \s-1FITNESS FOR A PARTICULAR PURPOSE.\s0  See the \s-1GNU\s0 General Public License for more details.


Changelog
Argonaut changelog
==================

* Argonaut 1.2

[Fix] Bugs #5461: Error with Ldap2zone slave when using multiple domains with a common reverse zone
[Fix] Bugs #5513: netboot stayed in uninstall state
[Fix] Bugs #5523: Periodical schedule is not repeated
[Fix] Bugs #5541: Errors when trying to schedule actions
[Fix] Bugs #5544: Periodical jobs miss the first launch
[Fix] Bugs #5546: OPSI update software stay in deployement queue with "in progress" status
[Feature] Bugs #5558: Use some trim on argonaut.conf
[Fix] Bugs #5559: Argonaut client and server systemd service definition fails
[Fix] Bugs #5646: changing the library used to send mail in argonaut-user-reminder

* Argonaut 1.1

[Feature] Bugs #5379: Argonaut should support fdMode instead of gotoMode for FD 1.1
[Feature] Bugs #5439: we should add the directive check-names into the argonaut-dns service
[Feature] Bugs #5440: argonaut-ldap2zone should be able to get data from a branch
[Feature] Bugs #5447: argonaut ldap2zone should be able to create config for dns slave also, but not create the data
[Feature] Wishlist #5332: argonaut-user-reminder should support ppolicy

* Argonaut 1.0

[Fix] Bugs #5273: Incompatibility between recovery password and user-reminder
[Feature] Bugs #5275: redesign the argonaut-fusiondirectory tools
[Feature] Bugs #5281: create a function in argonaut to read correctly the fusiondirectory config
[Fix] Bugs #5288: function branch_exists should go into common.pm
[Fix] Bugs #5300: Errors when I try to start argonaut-client

* Argonaut 0.9.8

[Feature] Bugs #5059: module for creating samba shares
[Fix] Bugs #5146: argonaut fuse is broken and doesnt add the initrd=initrd.img-install line to pxe configuration for fai
[Fix] Bugs #5147: removed man file created in pxelinux.cfg and replace it by a default file
[Fix] Bugs #5148: the default mode stored inside the argonaut-fuse should be used in argonaut-fuse
[Fix] Bugs #5158: FAI_ARGONAUT variable is not used inside the script
[Fix] Bugs #5159: we should add some debug to the ldap query in fai.pm
[Fix] Bugs #5164: argonaut-fai-monitor typo

* Argonaut 0.9.7

[Fix] Bugs #4691: the systemd unit files have a syntax error
[Feature] Bugs #4864: migrate the script for user-reminder to the argonaut-framework
[Feature] Bugs #4865: create argonaut-clean-audit to remove old audit entries from ldap

* Argonaut 0.9.6

[Feature] Bugs #4493: It would be nice to accept more than one reverse zone with argonaut-ldap2zone
[Feature] Bugs #4494: Add aaaa record in argonaut-ldap2zone
[Feature] Bugs #4503: Merge reverse zone with argonaut-ldap2zone
[Feature] Bugs #4506: Support for split horizon should be added
[Fix] Bugs #4513: Argonaut-ldap2zone have an perl error when we not have a reverse zone in the LDAP
[Feature] Bugs #4607: add a --ldap2acl option in ldap2zone

* Argonaut 0.9.5

[Fix] Bugs #4293: Problem when I try to restart a service on Centos7
[Fix] Bugs #4432: renomed and added the Service.systemd.pm and System.systemd.pm

* Argonaut 0.9.4

[Fix] Bugs #4110: Remove all uses of gotoBootKernel
[Fix] Bugs #4291: Error when I try to grab centos packages with an argonaut-server on centos
[Fix] Bugs #4293: Problem when I try to restart a service on Centos7
[Fix] Bugs #4322: generate-fusioninventory-schema should read Agent/Inventory.pm directly from an installed fusioninventory-agent
[Fix] Bugs #4355: Use App::Daemon for argonaut-fai-monitor and argonaut-fuse

* Argonaut 0.9.3

[Feature] Bugs #943: argonaut should work on ssl mode https
[Fix] Bugs #3791: Add TXT values in global zone record
[Feature] Bugs #3803: ldap2zone needs an option to not write reverse zone
[Feature] Bugs #3945: Add a option to not touch at the reverse zone
[Fix] Bugs #4046: I have an error in argonaut-fai-monitor log when I try to use it
[Feature] Bugs #4048: read the configuration from argonaut-fai-monitor service
[Feature] Bugs #4049: split argonaut-fai-monitor in his own package
[Fix] Bugs #4052: Argonaut-server must support that we send a MAC in uppercase
[Fix] Bugs #4166: update all manpages for 0.9.3

* Argonaut 0.9.2

[Feature] Bugs #2229: adding software on demand onto the opsi service
[Fix] Bugs #2454: gotoLdapServer seems unused
[Fix] Bugs #2577: Reboot with Opsi Client
[Fix] Bugs #2906: debconf2ldif is missing an help option
[Fix] Bugs #3140: adding debconf and freeradius to the argonaut 0.9.2 branch
[Fix] Bugs #3146: OPSI havewrong module place
[Fix] Bugs #3155: when I want to configure OPSI softwarelist with FD I got : Erreur: Request error: Error : No such a method : 'Argonaut.ClientDaemon.Modules.OPSI.get_localboots'.
[Fix] Bugs #3286: put comment around pod explanation in argonaut-server to correct error in manpage
[Fix] Bugs #3376: No client module can handle action Deployment.reinstall for target ff:ff:aa:aa:aa:aa
[Fix] Bugs #3382: quota plugins et block size
[Fix] Bugs #3390: rename ldap2fai into argonaut-ldap2fai to remove potential clash with the goto software from gonicus
[Fix] Bugs #3449: fusioninventory ldap schema generator is not in git anymore
[Feature] Bugs #3481: adding centos/rpm support to argonaut Packages.pm library
[Fix] Bugs #3526: add a switch to select the good library when in wheezy or jessie for argonaut-client
[Fix] Bugs #3541: fail2ldif manpage is wrong in the description
[Fix] Bugs #3542: when the parent serveur is on level up than the son for the fai repository it doesnt find it correctly
[Fix] Bugs #3543: when asking for the help of the fai2ldif command we get an extra h 1
[Feature] Bugs #3556: Support de Centos dans argonaut-common-fai
[Fix] Bugs #3558: argonaut-common-fai - Création de partition LVM
[Fix] Bugs #3574: fusioninventory ldap schema generator is not in git anymore
[Fix] Bugs #3576: clean and rename the freeradius argonaut code
[Fix] Bugs #3611: we should merge fai-monitor and argonaut-fai-monitor
[Fix] Bugs #3612: argonaut fuse module fai should check if the system is lock or not
[Fix] Bugs #3646: argonaut-fuse should be cleaned up

argonaut-server/Argonaut/Server/Modules/Argonaut.pm
#######################################################################
#
# Argonaut::Server::Modules::Argonaut -- Argonaut client module
#
# Copyright (C) 2012-2016 FusionDirectory project
#
# Authors: Côme BERNIGAUD
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
#
#######################################################################

package Argonaut::Server::Modules::Argonaut;

use strict;
use warnings;

use 5.008;

use Argonaut::Libraries::Common qw(:ldap :file :config :string);

use if (USE_LEGACY_JSON_RPC),     'JSON::RPC::Legacy::Client';
use if not (USE_LEGACY_JSON_RPC), 'JSON::RPC::Client';

my @unlocked_actions = ['System.halt', 'System.reboot'];

sub new
{
  my ($class) = @_;
  my $self = {};
  bless( $self, $class );
  return $self;
}

sub handle_client {
  my ($self, $mac,$action) = @_;

  $self->{target} = $mac;

  if ($action =~ m/^Deployment.(reboot|wake)$/) {
    $action =~ s/^Deployment./System./;
  } elsif ($action =~ m/^Deployment./) {
    $main::log->debug("[Argonaut] Can't handle Deployment actions");
    return 0;
  }

  if ($action eq 'System.wake') {
    $self->{action} = $action;
    return 1;
  }

  eval { #try
    my $settings = argonaut_get_client_settings($main::config,"(macAddress=$mac)");
    %$self = %$settings;
    $self->{action} = $action;
    $self->{target} = $mac;
  };
  if ($@) { #catch
    $main::log->debug("[Argonaut] Can't handle client : $@");
    return 0;
  };
  $self->{cacertfile} = $main::server_settings->{cacertfile};
  $self->{token}      = $main::server_settings->{token};
  # We take a lower timeout than the server so that it's possible to return the result
  $self->{timeout}    = $main::server_settings->{timeout} - 2;
  if ($self->{timeout} <= 0) {
    $self->{timeout} = 1;
  }

  return 1;
}

=pod
=item do_action
Execute a JSON-RPC method on a client which the ip is given.
Parameters : ip,action,params
=cut
sub do_action {
  my ($self, $params) = @_;
  my $action = $self->{action};

  if ($action eq 'System.wake') {
    main::wakeOnLan($self->{'mac'});
    return 1;
  }

  if ($self->{'locked'} && (grep {$_ eq $action} @unlocked_actions)) {
    die 'This computer is locked';
  }

  if ($action eq 'ping') {
    my $ok = 'OK';
    eval {
      my $res = $self->launch('echo',$ok);
      return ($res eq $ok);
    };
    return 0;
  } else {
    return $self->launch($action,$params);
  }
}

=pod
=item launch
Execute a JSON-RPC method on a client which the ip is given.
Parameters : ip,action,params
=cut
sub launch { # if ip pings, send the request
  my ($self, $action,$params) = @_;

  if ($action =~ m/^[^.]+\.[^.]+$/) {
    $action = 'Argonaut.ClientDaemon.Modules.'.$action;
  }

  my $ip = $self->{'ip'};
  # this line is only needed when debugging stuff on localhost
  #$ip = "localhost";

  $main::log->info("sending action $action to $ip");

  my $client;
  if (USE_LEGACY_JSON_RPC) {
    $client = new JSON::RPC::Legacy::Client;
  } else {
    $client = new JSON::RPC::Client;
  }
  $client->version('1.0');
  if ($self->{'protocol'} eq 'https') {
    if ($client->ua->can('ssl_opts')) {
      $client->ua->ssl_opts(
        verify_hostname   => 1,
        SSL_ca_file       => $self->{'cacertfile'},
        SSL_verifycn_name => $self->{'certcn'}
      );
      $client->ua->credentials($ip.":".$self->{'port'}, "JSONRPCRealm", "", argonaut_gen_ssha_token($self->{'token'}));
    }
  }
  $client->ua->timeout($self->{timeout});

  my $callobj = {
    method  => $action,
    params  => [$params],
  };

  my $res;
  eval {
    $res = $client->call($self->{'protocol'}."://".$ip.":".$self->{'port'}, $callobj);
  };
  if ($@) {
    if ($client->status_line =~ m/^(4|5)\d\d/) {
      $main::log->info("Status : ".$client->status_line);
      die $client->status_line."\n";
    } else {
      $main::log->error("Error : ".$@);
      die $@."\n";
    }
  }

  if($res) {
    if ($res->is_error) {
      $main::log->error("Error : ".$res->error_message);
      die "Error : ", $res->error_message."\n";
    }
    else {
      $main::log->info("Result : ".$res->result);
      return $res->result;
    }
  }
  else {
    $main::log->info("Status : ".$client->status_line);
    die "Status : ".$client->status_line."\n";
  }
}

1;

__END__
